web_poet/rules.py

```diff
@@ -0,0 +1,338 @@
+from __future__ import annotations  # https://www.python.org/dev/peps/pep-0563/
+
+import importlib
+import importlib.util
+import pkgutil
+import warnings
+from collections import deque
+from operator import attrgetter
+from typing import Any, Dict, Iterable, List, Optional, Type, TypeVar, Union
+
+import attrs
+from url_matcher import Patterns
+
+from web_poet._typing import get_item_cls
+from web_poet.pages import ItemPage
+from web_poet.utils import _create_deprecated_class, as_list, str_to_pattern
+
+Strings = Union[str, Iterable[str]]
+
+PageObjectRegistryTV = TypeVar("PageObjectRegistryTV", bound="PageObjectRegistry")
+
+
+@attrs.define(frozen=True)
+class ApplyRule:
+    """A rule that primarily applies Page Object and Item overrides for a given
+    URL pattern.
+
+    This is instantiated when using the :func:`web_poet.handle_urls` decorator.
+    It's also being returned as a ``List[ApplyRule]`` when calling the
+    ``web_poet.default_registry``'s :meth:`~.PageObjectRegistry.get_rules`
+    method.
+
+    You can access any of its attributes:
+
+        * ``for_patterns`` - contains the list of URL patterns associated with
+          this rule. You can read the API documentation of the `url-matcher
+          <https://url-matcher.readthedocs.io/>`_ package for more information
+          about the patterns.
+        * ``use`` - The Page Object that will be **used** in cases where the URL
+          pattern represented by the ``for_patterns`` attribute is matched.
+        * ``instead_of`` - *(optional)* The Page Object that will be **replaced**
+          with the Page Object specified via the ``use`` parameter.
+        * ``to_return`` - *(optional)* The data container class that will be
+          **replaced** by the item returned by the Page Object specified via the
+          ``use`` parameter.
+        * ``meta`` - *(optional)* Any other information you may want to store.
+          This doesn't do anything for now but may be useful for future API updates.
+
+    The main functionality of this class lies in the ``instead_of`` and ``to_return``
+    parameter. Should both of these be omitted, then :class:`~.ApplyRule` simply
+    tags which URL patterns the given Page Object is expected to be used.
```

Comment:
Not sure that's valid; if to_return is not present, then ApplyRule should be using ItemPage's return class.

Reply:
I see, we were assigning `to_return=None` before on cases like:

```python
@handle_urls("example.com")
class SomePage(ItemPage):
    ...
```

Updated this behavior in https://github.com/scrapinghub/web-poet/pull/84/commits/3679b59720e53fa74b7f58568310930b75418ab7

Reply:
Wait, I think in this case None is the right behavior.

Reply:
To clarify, you're thinking about the following case right? wherein the users haven't declared the Item class in both the Page Object and the Rule:

```python
@handle_urls("example.com")
class SomePage(ItemPage):
    ...
```

In this case, should `to_return` be assigned with `None` or `dict` as the default value?

In the https://github.com/scrapinghub/web-poet/commit/3679b59720e53fa74b7f58568310930b75418ab7 commit, I've changed it from `None` to `dict` since by default, the `Returns` class was using `dict` in a subtle way.

This means that ItemPage subclasses without any hints about what it's able to return is expected to return `dict`s by default.

Since we're also thinking about establishing a concept for Crawling POs, I think it could serve the purpose of having `to_return=None` since it's not really expected to return any time at all, but rather serve as a convenience object to extract links, paginations, etc used for crawling. On the other hand, Data POs should at least have `to_return=dict` as a default to establish the fact that it's capable of returning an item.

Reply:
I think we should special-case dict, because it doesn't make sense from the providers point of view. We'd have a lot of page objects which can return dicts. Does it mean that if user declares a dependency on a dict (or just annotates something as a dict), we should call some random page object?

We can special-case it in a library like scrapy-poet (and in other similar frameworks, when we have them). But it seems web-poet can be a better place to do it. I was thinking about `dict` as a fallback for the cases where a return item class is not defined, not as a real item class.

> Since we're also thinking about establishing a concept for Crawling POs, I think it could serve the purpose of having to_return=None since it's not really expected to return any item at all

I'm not sure about it. Crawling page objects could return an item with information needed for crawling - e.g. ProductList item for products. We will need to standardize these data types.

Reply:
I see, I think I understand what you mean now as I got it the other way around before. 😅 Updated in https://github.com/scrapinghub/web-poet/pull/84/commits/de15a86b8497b1dd40b29668a184409cc80f7177